### Importing libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Importing and Cleaning of the Dataset

In [2]:
data = pd.read_csv("Training.csv")

In [3]:
data.head()

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   
3        1          1                     0                    0          0   
4        1          1                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       0           0             0        0                 0  ...         0   
2       0           0             0        0                 0  ...         0   
3       0           0             0        0                 0  ...         0   
4       0           0             0        0                 0  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   
3             0                    0                     0   
4             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   
3                   0        0                     0                  0   
4                   0        0                     0                  0   

          prognosis  Unnamed: 133  
0  Fungal infection           NaN  
1  Fungal infection           NaN  
2  Fungal infection           NaN  
3  Fungal infection           NaN  
4  Fungal infection           NaN  

[5 rows x 134 columns]

In [4]:
## removing the last column of the dataframe
data.drop(data.columns[-1], axis=1,inplace=True)

In [5]:
data.head()

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   
3        1          1                     0                    0          0   
4        1          1                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   
1       0           0             0        0                 0  ...   
2       0           0             0        0                 0  ...   
3       0           0             0        0                 0  ...   
4       0           0             0        0                 0  ...   

   blackheads  scurring  skin_peeling  silver_like_dusting  \
0           0         0             0                    0   
1           0         0             0                    0   
2           0         0             0                    0   
3           0         0             0                    0   
4           0         0             0                    0   

   small_dents_in_nails  inflammatory_nails  blister  red_sore_around_nose  \
0                     0                   0        0                     0   
1                     0                   0        0                     0   
2                     0                   0        0                     0   
3                     0                   0        0                     0   
4                     0                   0        0                     0   

   yellow_crust_ooze         prognosis  
0                  0  Fungal infection  
1                  0  Fungal infection  
2                  0  Fungal infection  
3                  0  Fungal infection  
4                  0  Fungal infection  

[5 rows x 133 columns]

### Exploring the dataset

In [6]:
data.shape

(4920, 133)

In [7]:
data.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4920 entries, 0 to 4919
Columns: 133 entries, itching to prognosis
dtypes: int64(132), object(1)
memory usage: 5.3 MB


In [8]:
data.describe()

itching    skin_rash  nodal_skin_eruptions  continuous_sneezing  \
count  4920.000000  4920.000000           4920.000000          4920.000000   
mean      0.137805     0.159756              0.021951             0.045122   
std       0.344730     0.366417              0.146539             0.207593   
min       0.000000     0.000000              0.000000             0.000000   
25%       0.000000     0.000000              0.000000             0.000000   
50%       0.000000     0.000000              0.000000             0.000000   
75%       0.000000     0.000000              0.000000             0.000000   
max       1.000000     1.000000              1.000000             1.000000   

         shivering       chills   joint_pain  stomach_pain      acidity  \
count  4920.000000  4920.000000  4920.000000   4920.000000  4920.000000   
mean      0.021951     0.162195     0.139024      0.045122     0.045122   
std       0.146539     0.368667     0.346007      0.207593     0.207593   
min       0.000000     0.000000     0.000000      0.000000     0.000000   
25%       0.000000     0.000000     0.000000      0.000000     0.000000   
50%       0.000000     0.000000     0.000000      0.000000     0.000000   
75%       0.000000     0.000000     0.000000      0.000000     0.000000   
max       1.000000     1.000000     1.000000      1.000000     1.000000   

       ulcers_on_tongue  ...  pus_filled_pimples   blackheads     scurring  \
count       4920.000000  ...         4920.000000  4920.000000  4920.000000   
mean           0.021951  ...            0.021951     0.021951     0.021951   
std            0.146539  ...            0.146539     0.146539     0.146539   
min            0.000000  ...            0.000000     0.000000     0.000000   
25%            0.000000  ...            0.000000     0.000000     0.000000   
50%            0.000000  ...            0.000000     0.000000     0.000000   
75%            0.000000  ...            0.000000     0.000000     0.000000   
max            1.000000  ...            1.000000     1.000000     1.000000   

       skin_peeling  silver_like_dusting  small_dents_in_nails  \
count   4920.000000          4920.000000           4920.000000   
mean       0.023171             0.023171              0.023171   
std        0.150461             0.150461              0.150461   
min        0.000000             0.000000              0.000000   
25%        0.000000             0.000000              0.000000   
50%        0.000000             0.000000              0.000000   
75%        0.000000             0.000000              0.000000   
max        1.000000             1.000000              1.000000   

       inflammatory_nails      blister  red_sore_around_nose  \
count         4920.000000  4920.000000           4920.000000   
mean             0.023171     0.023171              0.023171   
std              0.150461     0.150461              0.150461   
min              0.000000     0.000000              0.000000   
25%              0.000000     0.000000              0.000000   
50%              0.000000     0.000000              0.000000   
75%              0.000000     0.000000              0.000000   
max              1.000000     1.000000              1.000000   

       yellow_crust_ooze  
count        4920.000000  
mean            0.023171  
std             0.150461  
min             0.000000  
25%             0.000000  
50%             0.000000  
75%             0.000000  
max             1.000000  

[8 rows x 132 columns]

In [9]:
data.isnull().any()

itching                 False
skin_rash               False
nodal_skin_eruptions    False
continuous_sneezing     False
shivering               False
                        ...  
inflammatory_nails      False
blister                 False
red_sore_around_nose    False
yellow_crust_ooze       False
prognosis               False
Length: 133, dtype: bool

- Decreasing memory usage by decreasing the size of different columns with int64

In [10]:
list_col_int64 = [col for col in data.columns if data[col].dtype == "int64"]
# a list to contain the names of all the columns of the dataset

In [11]:
for col in list_col_int64 :
    data[col] = pd.to_numeric(data[col], downcast="unsigned")

In [12]:
data.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4920 entries, 0 to 4919
Columns: 133 entries, itching to prognosis
dtypes: object(1), uint8(132)
memory usage: 971.3 KB


**********************************************************

- Memory usage :
- Before : 5.3 MB
- After : 971.3 Kb

### Data preprocessing

- Converting categorical data into numerical data

##### Dummy variable method

In [13]:
prognosis_dummies = pd.get_dummies(data.prognosis, prefix="Disease", drop_first="True")
prognosis_dummies.head()

Disease_AIDS  Disease_Acne  Disease_Alcoholic hepatitis  Disease_Allergy  \
0             0             0                            0                0   
1             0             0                            0                0   
2             0             0                            0                0   
3             0             0                            0                0   
4             0             0                            0                0   

   Disease_Arthritis  Disease_Bronchial Asthma  Disease_Cervical spondylosis  \
0                  0                         0                             0   
1                  0                         0                             0   
2                  0                         0                             0   
3                  0                         0                             0   
4                  0                         0                             0   

   Disease_Chicken pox  Disease_Chronic cholestasis  Disease_Common Cold  ...  \
0                    0                            0                    0  ...   
1                    0                            0                    0  ...   
2                    0                            0                    0  ...   
3                    0                            0                    0  ...   
4                    0                            0                    0  ...   

   Disease_Osteoarthristis  Disease_Paralysis (brain hemorrhage)  \
0                        0                                     0   
1                        0                                     0   
2                        0                                     0   
3                        0                                     0   
4                        0                                     0   

   Disease_Peptic ulcer diseae  Disease_Pneumonia  Disease_Psoriasis  \
0                            0                  0                  0   
1                            0                  0                  0   
2                            0                  0                  0   
3                            0                  0                  0   
4                            0                  0                  0   

   Disease_Tuberculosis  Disease_Typhoid  Disease_Urinary tract infection  \
0                     0                0                                0   
1                     0                0                                0   
2                     0                0                                0   
3                     0                0                                0   
4                     0                0                                0   

   Disease_Varicose veins  Disease_hepatitis A  
0                       0                    0  
1                       0                    0  
2                       0                    0  
3                       0                    0  
4                       0                    0  

[5 rows x 40 columns]

In [14]:
data_merged = pd.concat([data, prognosis_dummies], axis=1)
data_merged.head()

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   
3        1          1                     0                    0          0   
4        1          1                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   
1       0           0             0        0                 0  ...   
2       0           0             0        0                 0  ...   
3       0           0             0        0                 0  ...   
4       0           0             0        0                 0  ...   

   Disease_Osteoarthristis  Disease_Paralysis (brain hemorrhage)  \
0                        0                                     0   
1                        0                                     0   
2                        0                                     0   
3                        0                                     0   
4                        0                                     0   

   Disease_Peptic ulcer diseae  Disease_Pneumonia  Disease_Psoriasis  \
0                            0                  0                  0   
1                            0                  0                  0   
2                            0                  0                  0   
3                            0                  0                  0   
4                            0                  0                  0   

   Disease_Tuberculosis  Disease_Typhoid  Disease_Urinary tract infection  \
0                     0                0                                0   
1                     0                0                                0   
2                     0                0                                0   
3                     0                0                                0   
4                     0                0                                0   

   Disease_Varicose veins  Disease_hepatitis A  
0                       0                    0  
1                       0                    0  
2                       0                    0  
3                       0                    0  
4                       0                    0  

[5 rows x 173 columns]

In [15]:
data_merged.shape

(4920, 173)

In [16]:
df = data_merged.drop("prognosis", axis=1)
df.shape

(4920, 172)

##### Label encoding method

In [17]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [18]:
data_le = pd.DataFrame(le.fit_transform(data[["prognosis"]]), columns=["Disease"])

E:\Anaconda\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [19]:
data_le

Disease
0          15
1          15
2          15
3          15
4          15
...       ...
4915        0
4916        2
4917       38
4918       35
4919       27

[4920 rows x 1 columns]

In [20]:
merging_data = pd.concat([data, data_le], axis=1)
merging_data.head(3)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       0           0             0        0                 0  ...         0   
2       0           0             0        0                 0  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   

          prognosis  Disease  
0  Fungal infection       15  
1  Fungal infection       15  
2  Fungal infection       15  

[3 rows x 134 columns]

In [21]:
disease_prognosis = pd.DataFrame(merging_data.prognosis)
disease_prognosis["Disease"] = merging_data.Disease
# disease_prognosis.drop_duplicates()
disease_prognosis.head(3)

prognosis  Disease
0  Fungal infection       15
1  Fungal infection       15
2  Fungal infection       15

In [22]:
disease_prognosis.shape

(4920, 2)

In [23]:
disease_prognosis = disease_prognosis.drop_duplicates().reset_index()
disease_prognosis.drop("index", axis=1, inplace=True)
disease_prognosis.head(3)

prognosis  Disease
0  Fungal infection       15
1           Allergy        4
2              GERD       16

In [24]:
disease_dict = {}

In [25]:
# final_data = merging_data.drop("prognosis",axis=1)
# you can remove the columns later

In [26]:
merging_data.head(3)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       0           0             0        0                 0  ...         0   
2       0           0             0        0                 0  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   

          prognosis  Disease  
0  Fungal infection       15  
1  Fungal infection       15  
2  Fungal infection       15  

[3 rows x 134 columns]

In [27]:
X_train = merging_data.drop(["prognosis", "Disease"], axis=1)
y_train = merging_data.Disease

### Model creation

In [28]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

In [29]:
lr.fit(X_train,y_train)

LogisticRegression()

In [30]:
test_data = pd.read_csv("Testing.csv")
test_data.head(3)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          0                     0                    1          1   
2        0          0                     0                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   
1       1           0             0        0                 0  ...   
2       0           0             1        1                 1  ...   

   blackheads  scurring  skin_peeling  silver_like_dusting  \
0           0         0             0                    0   
1           0         0             0                    0   
2           0         0             0                    0   

   small_dents_in_nails  inflammatory_nails  blister  red_sore_around_nose  \
0                     0                   0        0                     0   
1                     0                   0        0                     0   
2                     0                   0        0                     0   

   yellow_crust_ooze         prognosis  
0                  0  Fungal infection  
1                  0           Allergy  
2                  0              GERD  

[3 rows x 133 columns]

In [31]:
test_data["Disease"] = le.fit_transform(test_data.prognosis)
test_data.head(3)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          0                     0                    1          1   
2        0          0                     0                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       1           0             0        0                 0  ...         0   
2       0           0             1        1                 1  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   

          prognosis  Disease  
0  Fungal infection       15  
1           Allergy        4  
2              GERD       16  

[3 rows x 134 columns]

In [32]:
X_test = test_data.drop(["Disease", "prognosis"], axis=1)
y_test = test_data.Disease

In [33]:
X_test.shape

(42, 132)

In [34]:
lr.score(X_test, y_test)

1.0

#### Algorithm : Logistic Regression
#### Model accuracy : 1.0